File: LeetCode/ArrayStrings/python/fullTextJustify.py
from typing import List


class Solution:
    def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:
        res = []
        line = []
        length = 0
        i = 0

        while i < len(words):
            if length + len(line) + len(words[i]) > maxWidth:
                extra_space = maxWidth - length
                spaces = extra_space // max(1, len(line) - 1)
                remainder = extra_space % max(1, len(line) - 1)

                for j in range(0, max(1, len(line) - 1)):
                    line[j] += " " * spaces

                    if remainder:
                        line[j] += " "
                        remainder -= 1
                
                res.append("".join(line))
                line = []
                length = 0
            
            line.append(words[i])
            length += len(words[i])
            i += 1

        last_line = " ".join(line)
        trail_space = maxWidth - len(last_line)
        res.append(last_line + " " * trail_space)
        
        
        return res

    @staticmethod
    def testSolution(input) -> None:
        print("Input: words {}".format(input["words"]))
        print("Input: maxWidth {}".format(input["maxWidth"]))
        print("Expected: {}".format(input["expected"]))
        val = Solution().fullJustify(input["words"], input["maxWidth"])
        print("Result: {}, {}".format(val,("Correct" if "".join(val) == "".join(input["expected"]) else "Wrong")))
        print("-" * 50)

if __name__ == "__main__":
    input = {"words": ["This", "is", "an", "example", "of", "text", "justification."] ,
             "maxWidth": 16, 
             "expected": ["This    is    an","example  of text","justification.  "]}
    Solution.testSolution(input)

    input = {"words": ["What","must","be","acknowledgment","shall","be"] ,
             "maxWidth": 16, 
             "expected": ["What   must   be","acknowledgment  ","shall be        "]}
    Solution.testSolution(input)

    input = {"words": ["Science","is","what","we","understand","well","enough","to","explain","to","a","computer.","Art","is","everything","else","we","do"] ,
             "maxWidth": 20, 
             "expected": ["Science  is  what we","understand      well","enough to explain to","a  computer.  Art is","everything  else  we","do                  "]}
    Solution.testSolution(input)

File: LeetCode/ArrayStrings/java/fullTextJustify.java
package LeetCode.ArrayStrings.java;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

record fullTextJustifyRecord(String[] words, int maxWidth, List<String> expected) {
}

public class fullTextJustify {
    public List<String> fullJustify(String[] words, int maxWidth) {

        List<String> res = new ArrayList<>();
        List<StringBuilder> line = new ArrayList<>();
        int length = 0;
        int i = 0;

        while (i < words.length) {
            if (length + line.size() + words[i].length() > maxWidth) {
                int extra_space = maxWidth - length;
                int spaces = (extra_space / Math.max(1, line.size() - 1));
                int remainder = extra_space % Math.max(1, line.size() - 1);

                for (int j = 0; j < Math.max(1, line.size() - 1); j += 1) {

                    line.get(j).append(" ".repeat(spaces));

                    if (remainder > 0) {
                        line.get(j).append(" ");
                        remainder -= 1;
                    }
                }

                res.add(String.join("", line));
                line.clear();
                length = 0;
            }

            line.add(new StringBuilder().append(words[i]));

            length += words[i].length();
            i += 1;
        }

        String last_line = String.join(" ", line);
        int trail_space = maxWidth - last_line.length();
        res.add(last_line + " ".repeat(trail_space));

        return res;
    }

    public static void main(String[] args) {
        fullTextJustifyRecord input = new fullTextJustifyRecord(
                new String[] { "This", "is", "an", "example", "of", "text", "justification." },
                16,
                List.of("This    is    an", "example  of text", "justification.  "));

        fullTextJustify.testSolution(input);

        input = new fullTextJustifyRecord(
                new String[] { "What", "must", "be", "acknowledgment", "shall", "be" },
                16,
                List.of("What   must   be", "acknowledgment  ", "shall be        "));

        fullTextJustify.testSolution(input);

        input = new fullTextJustifyRecord(
                new String[] { "Science", "is", "what", "we", "understand", "well", "enough", "to", "explain", "to",
                        "a", "computer.", "Art", "is", "everything", "else", "we",
                        "do" },
                20,
                List.of("Science  is  what we", "understand      well", "enough to explain to", "a  computer.  Art is",
                        "everything  else  we", "do                  "));

        fullTextJustify.testSolution(input);
    }

    private static void testSolution(fullTextJustifyRecord input) {
        System.out.println("Input: words: " + Arrays.toString(input.words()));
        System.out.println("Input: maxWidth: " + input.maxWidth());
        System.out.println("Expected: \t" + input.expected().toString());
        List<String> val = new fullTextJustify().fullJustify(input.words(), input.maxWidth());
        System.out.println("Result: \t" + val +
                "\nStatus: " + (val.equals(input.expected()) ? "Correct" : "Wrong"));
        System.out.println("-".repeat(50));
    }
}
